### PR TITLE
upcoming: [DI-21812] - Unit tests for widgets

### PR DESCRIPTION
**Pull request by ankita-akamai**
## Description 📝

Added unit tests for widgets.

## Changes  🔄

List any change(s) relevant to the reviewer.

- 
- ...

## Target release date 🗓️

Please specify a release date (and environment, if applicable) to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷
--

### Verification steps

(How to verify changes)

- [ ] ...
- [ ] ...

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

*Check all that apply*

- [ ] Use React components instead of HTML Tags
- [ ] Proper naming conventions like cameCase for variables & Function & snake_case for constants
- [ ] Use appropriate types & avoid using "any"
- [ ] No type casting & non-null assertions
- [ ] Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] Providing/Improving test coverage
- [ ] Use sx props to pass styles instead of style prop
- [ ] Add JSDoc comments for interface properties & functions
- [ ] Use strict equality (===) instead of double equal (==)
- [ ] Use of named arguments (interfaces) if function argument list exceeds size 2
- [ ] Destructure the props
- [ ] Keep component size small & move big computing functions to separate utility
- [ ] 📱 Providing mobile support

<br/>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

</details>

---

### CODE REVIEW

File: packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  generateGraphData,
+  generateMaxUnit,
+  getDimensionName,
+  getLabelName,
+  mapResourceIdToName,
+} from './CloudPulseWidgetUtils';
+
+import type { CloudPulseMetricsResponse } from '@linode/api-v4';
+
+it('test generateMaxUnit method', () => {
```
Comment (nikhagra-akamai) [inline packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts]: Add a test case for some unit which could be rolled up like data is in bytes & maximum possible roll up could be MB, KB etc. Here % percentage we are not supporting in roll up so there will be not any maximum rolled up unit

**Review by nikhagra-akamai — CHANGES_REQUESTED** [review]

File: packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        max: 2500000,
+        total: 6000000,
+      },
+      format: formatPercentage,
+      legendColor: '#000',
+      legendTitle: 'linode-2',
+    },
+  ];
+
+  const result = generateMaxUnit(legendRowsData, 'Bytes');
```
Comment (venkymano-akamai) [inline packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts]: Overall, this is a minimalistic test method and does not cover edge cases. Please add some cases like empty array etc.,

File: packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    label: 'CPU Usage',
+    metric: { resource_id: '123' },
+    resources: [{ id: '123', label: 'linode-1' }],
+    serviceType: 'linode',
+    unit: '%',
+  };
+
+  const result = getLabelName(props);
+  expect(result).toBe('linode-1');
```
Comment (venkymano-akamai) [inline packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts]: same here, can we think all angles and possible edge cases, if we cover that we can uncover more issues in the code and avoid UI breaking also if we are lucky

File: packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+vi.mock('../Utils/UserPreference', () => ({
+  useAclpPreference: () => ({
+    updateWidgetPreference: mockUpdatePreferences,
+  }),
+}));
+
+describe('Cloud pulse widgets', () => {
+  window.ResizeObserver = ResizeObserver;
+  it('should render the widget with correct title and unit', () => {
+    const { getByText } = renderWithTheme(<CloudPulseWidget {...props} />);
```
Comment (venkymano-akamai) [inline packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx]: Since this is a single widget, you can validate the widget title, the filters like aggregation and interval filter everything here itself

File: packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      size: 6,
+    });
+  });
+
+  it('should update preferences for aggregation function select', async () => {
+    const user = userEvent.setup();
+
+    const { getByRole } = renderWithTheme(<CloudPulseWidget {...props} />);
+
+    await user.click(
```
Comment (venkymano-akamai) [inline packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx]: we can just use userEvent.click right?

File: packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+      getByRole('combobox', { name: 'Select an Aggregate Function' })
+    );
+
+    await user.click(getByRole('option', { name: 'Max' }));
+
+    expect(mockUpdatePreferences).toHaveBeenCalledWith('CPU Utilization', {
```
Comment (venkymano-akamai) [inline packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx:165]: Like this can we add validation for api request as well, like metrics api is called with required params?

**Review by venkymano-akamai — CHANGES_REQUESTED** [review]
@ankita-akamai , may be the util related tests add more edge cases, since we can cover all possibilities there

File: packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx (hunk 5 above)
Comment (ankita-akamai) [reply packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx]: I went through some docs, setup seems to be recommended - https://github.com/testing-library/user-event/discussions/1036, https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent. But, there is no big diff., so userEvent.click can also be used. For less code, i can replace with the userEvent.click.

**Review by nikhagra-akamai — APPROVED** [review]

File: packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts (hunk 2 above)
Comment (venkymano-akamai) [reply packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts]: @ankita-akamai , is this done?

**Review by venkymano-akamai — APPROVED** [review]
@ankita-akamai , approving pending one comment

File: packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts (hunk 2 above)
Comment (ankita-akamai) [reply packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.test.ts]: Done, as of now, I've enhanced it, but this func is little confusing, has redundant checks for undefined cases within it even after having defined types..
